### PR TITLE
follow,unfollow

### PR DESCRIPTION
**Pull request by thuytathi**

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -1,8 +1,9 @@
 class UsersController < ApplicationController
+  before_action :logged_in_user, only: [:index, :edit, :update, :destroy,
+                                        :following, :followers]
   before_action :logged_in_user, only: [:index, :edit, :update, :destroy]
```
Comment (dieunb) [inline app/controllers/users_controller.rb]: Duplicate before_action :logged_in_user

File: config/routes.rb [hunk 2]
```diff
@@ -20,10 +18,15 @@
   get 'contact' => 'static_pages#contact'
 
   root 'static_pages#home'
-
+    resources :users do
```
Comment (dieunb) [inline config/routes.rb]: Indent size = 2

File: config/routes.rb [hunk 3]
```diff
@@ -20,10 +18,15 @@
...
+    member do
+      get :following, :followers
+    end
+  end
   resources :users
```
Comment (dieunb) [inline config/routes.rb]: duplicated resources :users

**Comment by dieunb** [comment]
OK

File: app/controllers/relationships_controller.rb [hunk 4]
```diff
@@ -0,0 +1,20 @@
+class RelationshipsController < ApplicationController
+  before_action :logged_in_user
+  def create
+    @user = User.find(params[:followed_id])
```
Comment (chuanhtuan) [inline app/controllers/relationships_controller.rb]: ()

File: app/controllers/relationships_controller.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    current_user.follow(@user)
```
Comment (chuanhtuan) [inline app/controllers/relationships_controller.rb]: ()

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -47,12 +47,27 @@ def destroy
     redirect_to users_url
   end
 
-   private
+  def following
+    @title = "Following"
+    @user  = User.find(params[:id])
```
Comment (chuanhtuan) [inline app/controllers/users_controller.rb]: ()

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -47,12 +47,27 @@ def destroy
...
+    @users = @user.following.paginate(page: params[:page])
```
Comment (chuanhtuan) [inline app/controllers/users_controller.rb]: ()

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -47,12 +47,27 @@ def destroy
...
+    render 'show_follow'
```
Comment (chuanhtuan) [inline app/controllers/users_controller.rb]: use "

File: app/controllers/users_controller.rb [hunk 9]
```diff
@@ -47,12 +47,27 @@ def destroy
...
+  end
+
+  def followers
+    @title = "Followers"
+    @user  = User.find(params[:id])
```
Comment (chuanhtuan) [inline app/controllers/users_controller.rb]: ()

File: app/views/users/show.html.erb [hunk 10]
```diff
@@ -1,11 +1,17 @@
...
   <aside class="col-md-4">
-    <section class="user_info">
+    <section>
       <h1>
         <%= gravatar_for @user %>
         <%= @user.name %>
       </h1>
     </section>
+    <section class="stats">
+      <%= render 'shared/stats' %>
```
Comment (chuanhtuan) [inline app/views/users/show.html.erb]: "

File: config/routes.rb [hunk 11]
```diff
@@ -20,49 +18,52 @@
   get 'contact' => 'static_pages#contact'
 
   root 'static_pages#home'
-
-  resources :users
+  resources :users do
+    member do
+      get :following, :followers
```
Comment (chuanhtuan) [inline config/routes.rb]: remove

File: app/controllers/application_controller.rb [hunk 12]
```diff
@@ -4,4 +4,13 @@ class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
   include SessionsHelper
 
+  private
+  # Confirms a logged-in user.
```
Comment (dieunb) [inline app/controllers/application_controller.rb]: Remove comment code `#Confirms a logged-in user.`

File: app/controllers/users_controller.rb [hunk 13]
```diff
@@ -62,14 +77,15 @@ def logged_in_user
...
- # Confirms the correct user.
+  # Confirms the correct user.
   def correct_user
-    @user = User.find(params[:id])
-    redirect_to(root_url) unless current_user?(@user)
+    @user = User.find params[:id]
+    redirect_to root_url unless current_user? @user
   end
+
   # Confirms an admin user.
```
Comment (dieunb) [inline app/controllers/users_controller.rb]: remove `# Confirms an admin user.`

File: app/controllers/users_controller.rb [hunk 14]
```diff
@@ -47,12 +47,27 @@ def destroy
...
+
+  private
 
   def user_params
     params.require(:user).permit(:name, :email, :password,
-                                 :password_confirmation)
+    :password_confirmation)
   end
+
   # Confirms a logged-in user.
```
Comment (dieunb) [inline app/controllers/users_controller.rb]: remove `# Confirms a logged-in user.`

File: app/models/user.rb [hunk 15]
```diff
@@ -70,6 +76,20 @@ def send_password_reset_email
   def password_reset_expired?
     reset_sent_at < 2.hours.ago
   end
+  # Follows a user.
```
Comment (dieunb) [inline app/models/user.rb]: remove `# Follows a user.`

File: app/models/user.rb [hunk 16]
```diff
@@ -70,6 +76,20 @@ def send_password_reset_email
...
+  def follow(other_user)
+    active_relationships.create(followed_id: other_user.id)
+  end
+
+  # Unfollows a user.
```
Comment (dieunb) [inline app/models/user.rb]: remove `# Unfollows a user.`

File: app/views/users/_unfollow.html.erb [hunk 17]
```diff
@@ -0,0 +1,4 @@
+<%= form_for(current_user.active_relationships.find_by(followed_id: @user.id),
```
Comment (dieunb) [inline app/views/users/_unfollow.html.erb]: Don't put logic to view, move `current_user.active_relationships.find_by(followed_id: @user.id)` into model or controller and remember don't use `()`

File: app/controllers/users_controller.rb [hunk 18]
```diff
@@ -47,13 +47,27 @@ def destroy
     redirect_to users_url
   end
 
-   private
+  def following
```
Comment (chuanhtuan) [inline app/controllers/users_controller.rb]: RESTful :(

File: app/models/user.rb [hunk 19]
```diff
@@ -14,63 +20,73 @@ class User < ActiveRecord::Base
                     uniqueness: { case_sensitive: false }
   has_secure_password
   validates :password, length: { minimum: 6 }
-  # Activates an account.
+
   def activate
     update_attribute(:activated,    true)
```
Comment (dieunb) [inline app/models/user.rb]: group all to `update_attributes(:activated: true, activated_at: Time.now)` 
`Time.zone.now` => wrong

File: app/models/user.rb [hunk 20]
```diff
@@ -14,63 +20,73 @@ class User < ActiveRecord::Base
...
   end
-  # Forgets a user.
+  
   def forget
     update_attribute(:remember_digest, nil)
-  end
-  # Creates and assigns the activation token and digest.
+  end  
   def create_activation_digest
       self.activation_token  = User.new_token
```
Comment (dieunb) [inline app/models/user.rb]: indent

File: app/models/user.rb [hunk 21]
```diff
@@ -14,63 +20,73 @@ class User < ActiveRecord::Base
...
   def send_password_reset_email
     UserMailer.password_reset(self).deliver_now
   end
-  # Returns true if a password reset has expired.
   def password_reset_expired?
     reset_sent_at < 2.hours.ago
   end
 
+  def follow(other_user)
+    active_relationships.create(followed_id: other_user.id)
```
Comment (dieunb) [inline app/models/user.rb]: use 
```
self.followeds << other_user
self.save
```
instead of it

File: app/models/user.rb [hunk 22]
```diff
@@ -14,63 +20,72 @@ class User < ActiveRecord::Base
                     uniqueness: { case_sensitive: false }
   has_secure_password
   validates :password, length: { minimum: 6 }
-  # Activates an account.
+
   def activate
-    update_attribute(:activated,    true)
-    update_attribute(:activated_at, Time.zone.now)
+    update_attributes(:activated => true,:activated_at => Time.now)
```
Comment (dieunb) [inline app/models/user.rb]: `update_attributes(activated: true, activated_at: Time.now)`

File: app/models/user.rb [hunk 23]
```diff
@@ -14,63 +20,72 @@ class User < ActiveRecord::Base
...
   def send_password_reset_email
     UserMailer.password_reset(self).deliver_now
   end
-  # Returns true if a password reset has expired.
   def password_reset_expired?
     reset_sent_at < 2.hours.ago
   end
 
+  def follow(other_user)
+    self.following << other_user
```
Comment (dieunb) [inline app/models/user.rb]: `self.followings`

File: app/models/user.rb [hunk 24]
```diff
@@ -14,63 +20,72 @@ class User < ActiveRecord::Base
...
+    following.include?(other_user)
+  end
+  
+  def relation_with(user)
+    self.active_relationships.find_by followed: user
+  end
+
  private
   def downcase_email
     self.email=email.downcase
```
Comment (dieunb) [inline app/models/user.rb]: `add space before and after =`

File: app/models/user.rb [hunk 25]
```diff
@@ -79,6 +94,4 @@ def create_activation_digest
     self.activation_token=User.new_token
```
Comment (dieunb) [inline app/models/user.rb]: add space before and after `=`

File: app/models/user.rb [hunk 26]
```diff
@@ -79,6 +94,4 @@ def create_activation_digest
...
     self.activation_digest=User.digest(activation_token)
```
Comment (dieunb) [inline app/models/user.rb]: add space before and after `=`

File: app/views/shared/_stats.html.erb [hunk 27]
```diff
@@ -0,0 +1,15 @@
+<% @user ||= current_user %>
+<div class="stats">
+  <a href="<%= user_path(@user)+following_path %>">
```
Comment (dieunb) [inline app/views/shared/_stats.html.erb]: `user_path(@user)+following_path` link?

File: app/views/shared/_stats.html.erb [hunk 28]
```diff
@@ -0,0 +1,15 @@
...
+    <strong id="following" class="stat">
+      <%= @user.following.count %>
+    </strong>
+    following
+  </a>
+  <a href="<%= user_path(@user)+followers_path %>">
```
Comment (dieunb) [inline app/views/shared/_stats.html.erb]: link? ` user_path(@user)+followers_path`

**Pull request merged**